common/src/main/scala/uk/ac/wellcome/dynamo/VersionedDao.scala

```diff
@@ -2,57 +2,164 @@ package uk.ac.wellcome.dynamo
 
 import com.amazonaws.services.dynamodbv2.AmazonDynamoDB
 import com.google.inject.Inject
+import com.gu.scanamo.query.{KeyEquals, UniqueKey}
 import com.gu.scanamo.syntax.{attributeExists, not, _}
+import com.gu.scanamo.update.UpdateExpression
 import com.gu.scanamo.{DynamoFormat, Scanamo, Table}
 import com.twitter.inject.Logging
+import shapeless.{HList, LabelledGeneric, Poly1, Witness}
+import shapeless.labelled.FieldType
 import uk.ac.wellcome.models._
+import shapeless._
+import shapeless.ops.hlist.{LeftFolder, Mapper}
+import shapeless.ops.record.Remover
+import shapeless.ops.tuple.ToTraversable
+import shapeless.record._
+import uk.ac.wellcome.dynamo.UpdateExpressionGenerator.toUpdateExpressions
 import uk.ac.wellcome.models.aws.DynamoConfig
+import uk.ac.wellcome.type_classes.IdGetter.{createIdGetter, IdKey}
+import uk.ac.wellcome.type_classes.{IdGetter, VersionGetter, VersionUpdater}
 import uk.ac.wellcome.utils.GlobalExecutionContext.context
 
+import scala.collection.generic.CanBuildFrom
 import scala.concurrent.Future
 
+trait UpdateExpressionGenerator[T] {
+  def generateUpdateExpression(t: T): Option[UpdateExpression]
+}
+
+object UpdateExpressionGenerator {
+  val w = Witness(Symbol("id"))
+  type IdKey = w.T
+  type IdField = FieldType[IdKey, String]
+
+  object toUpdateExpressions extends Poly1 {
+    implicit def some[K <: Symbol, V](implicit witness: Witness.Aux[K],
+                                      dynamoFormat: DynamoFormat[V])
+      : Case.Aux[FieldType[K, V], UpdateExpression] = {
+
+      at[FieldType[K, V]] { fieldtype =>
+        {
+          val fieldValue: V = fieldtype
+          set(witness.value -> fieldValue)
+        }
+      }
+    }
+  }
+
+  object Folder extends Poly2 {
```

Comment:
And this does a left fold over update expressions? So we can combine them into one?

Why is this extending Poly2 instead of Poly1? Is there a difference?